### PR TITLE
[Merged by Bors] - feat(Tactic/CategoryTheory): normalization in monoidal categories

### PR DESCRIPTION
**Pull request by yuma-mizuno**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+  | NormalExpr.nil _ => []
+  | NormalExpr.cons _ η ηs => η :: NormalExpr.toList ηs
+
+end Mathlib.Tactic.Monoidal
+
+open Mathlib.Tactic.Monoidal
+
+/-- `normalize% η` is the normalization of the 2-morphism `η`. It is of the form
+`α₀ ≫ η₀ ≫ α₁ ≫ η₁ ≫ ... αₙ ≫ ηₙ ≫ αₙ₊₁`, where `αᵢ` are structural 2-morphisms
+and `ηᵢ` are non-structural 2-morphisms. -/
```
Comment (Vtec234) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:658]: This description of normal forms seems a bit less specific than the description in the module doc at the top. Does the more specific description apply here?

**Comment by Vtec234** [comment]
What is the status of this PR? Is it still a draft or could be marked `awaiting-review`?

**Comment by yuma-mizuno** [comment]
@Vtec234 Thank you for the review! I think it's ready for review. I've opened the PR.

**Review by BoltonBailey — APPROVED** [review]
I looked over this, and nothing jumped out, though I am not super experienced with metaprogramming code.

Perhaps @yuma-mizuno could either resolve the above conversation or mark this PR as awaiting author?

**Comment by yuma-mizuno** [comment]
I apologize for the late response. I updated the docs following the suggestion by @Vtec234's suggestion.

I also added a note about what to do if proof term generation fails. The failure to generate proof terms occurs when using the normalization function in the string diagram widget. As a workaround we generate a meaningless term instead of generating an error.

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+This file provides a tactic that normalizes morphisms in monoidal categories. This is used in the
+string diagram widget given in `Mathlib.Tactic.StringDiagram`.
+We say that the morphism `η` in a monoidal category is in normal form if
+1. `η` is of the form `α₀ ≫ η₀ ≫ α₁ ≫ η₁ ≫ ... αₘ ≫ ηₘ ≫ αₘ₊₁` where each `αᵢ` is a
+  structural 2-morphism (consisting of associators and unitors),
+2. each `ηᵢ` is a non-structural 2-morphism of the form `f₁ ◁ ... ◁ fₘ ◁ θ`, and
+3. `θ` is of the form `ι ▷ g₁ ▷ ... ▷ gₗ`
+
+Note that the structural morphisms `αᵢ` are not necessarily normalized, as the main purpose
+to be to get a list of the non-structural morphisms out.
```
Comment (semorrison) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:19]:
```suggestion
is to get a list of the non-structural morphisms out.
```

**Comment by semorrison** [comment]
Could you write an account somewhere of your plans with this (or point to existing Zulip discussion etc)?

I'm excited to have widget support for string diagrams, but I would like to know that
* this code will be maintainable into the future
* its learnt whatever lessons necessary from e.g. homotopy.io

and I think a bit more documentation would help produce this confidence!

**Comment by yuma-mizuno** [comment]
@semorrison Thanks for the comment! I have added the note explaining the current situation and the future plan.

I've played a bit with homotopy.io before, but to be honest I don't know much about it (But now I should learn it).

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+the concept of strict monoidal categories due to the feature of dependent type theory. The
+normalization tactic can remove associators and unitors from the expression, extracting the
+necessary data for drawing string diagrams.
+
+The current plan on drawing string diagrams (#10581) is to use
+Penrose (https://github.com/penrose) via ProofWidget. However, it should be noted that the
+normalization procedure in this file does not rely on specific settings, allowing for broader
+application.
+
+Future plans include the following. At least I (Yuma) would like to work on tehse in the future,
```
Comment (semorrison) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:34]:
```suggestion
Future plans include the following. At least I (Yuma) would like to work on these in the future,
```

File: Mathlib/Tactic/CategoryTheory/Monoidal.lean [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+    let ⟨ι, pf_ι⟩ ← evalWhiskerLeftExpr f θ
+    let h := η.src
+    let h' := η.tar
+    let ⟨ι', pf_ι'⟩ ← evalComp ι (NormalExpr.associatorInv f g h')
+    let ⟨ι'', pf_ι''⟩ ← evalComp (NormalExpr.associator f g h) ι'
+    try return ⟨ι'', ← mkAppM ``evalWhiskerLeft_comp #[pf_θ, pf_ι, pf_ι', pf_ι'']⟩
+    catch _ => return ⟨ι'', mkConst ``True⟩
+  | .id, η => do
+    let f := η.src
+    let g := η.tar
```
Comment (semorrison) [inline Mathlib/Tactic/CategoryTheory/Monoidal.lean:602]: I think `tgt` is a better abbreviation.

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**